File: deegree2-igeoportal/src/main/webapp/javascript/geotransform.js
//$HeadURL$
/*----------------------------------------------------------------------------
 This file is part of deegree, http://deegree.org/
 Copyright (C) 2001-2009 by:
 - Department of Geography, University of Bonn -
 and
 - lat/lon GmbH -

 This library is free software; you can redistribute it and/or modify it under
 the terms of the GNU Lesser General Public License as published by the Free
 Software Foundation; either version 2.1 of the License, or (at your option)
 any later version.
 This library is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 details.
 You should have received a copy of the GNU Lesser General Public License
 along with this library; if not, write to the Free Software Foundation, Inc.,
 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

 Contact information:

 lat/lon GmbH
 Aennchenstr. 19, 53177 Bonn
 Germany
 http://lat-lon.de/

 Department of Geography, University of Bonn
 Prof. Dr. Klaus Greve
 Postfach 1147, 53001 Bonn
 Germany
 http://www.geographie.uni-bonn.de/deegree/

 e-mail: info@deegree.org
----------------------------------------------------------------------------*/

function GeoTransform(srcx1, srcy1, srcx2, srcy2, destx1, desty1, destx2, desty2) {
    this.qx = 0.0;
    this.qy = 0.0;
    
    this.srcx1 = srcx1;
    this.srcy1 = srcy1;
    this.srcx2 = srcx2;
    this.srcy2 = srcy2;
    this.destx1 = destx1;
    this.desty1 = desty1;
    this.destx2 = destx2;
    this.desty2 = desty2;

    this.getDestX = getDestX;
    this.getDestY = getDestY;
    this.getSourceX = getSourceX;
    this.getSourceY = getSourceY;
    this.calculateQX = calculateQX;
    this.calculateQY = calculateQY;
    this.calculateQX();
    this.calculateQY();
        
}

function getDestX( srcx ) {
    return this.destx1 + (srcx - this.srcx1) * this.qx;
}

function getDestY( srcy ) {
    return this.desty1 + (this.desty2-this.desty1) - (srcy - this.srcy1) * this.qy;
}

function getSourceX( destx ) {  	  
    return (destx - this.destx1) / this.qx - (-this.srcx1);
}

function getSourceY( desty ) {
    return ( (this.desty2 - this.desty1) - (desty - this.desty1) ) / this.qy - (- this.srcy1);
}

function calculateQX() {
    this.qx = (this.destx2-this.destx1)/(this.srcx2-this.srcx1);
}
    
function calculateQY() {
    this.qy = (this.desty2-this.desty1)/(this.srcy2-this.srcy1);
}

File: deegree2-core/src/main/java/org/deegree/graphics/transformation/GeoTransform.java
//$HeadURL$
/*----------------------------------------------------------------------------
 This file is part of deegree, http://deegree.org/
 Copyright (C) 2001-2009 by:
   Department of Geography, University of Bonn
 and
   lat/lon GmbH

 This library is free software; you can redistribute it and/or modify it under
 the terms of the GNU Lesser General Public License as published by the Free
 Software Foundation; either version 2.1 of the License, or (at your option)
 any later version.
 This library is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 details.
 You should have received a copy of the GNU Lesser General Public License
 along with this library; if not, write to the Free Software Foundation, Inc.,
 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

 Contact information:

 lat/lon GmbH
 Aennchenstr. 19, 53177 Bonn
 Germany
 http://lat-lon.de/

 Department of Geography, University of Bonn
 Prof. Dr. Klaus Greve
 Postfach 1147, 53001 Bonn
 Germany
 http://www.geographie.uni-bonn.de/deegree/

 e-mail: info@deegree.org
----------------------------------------------------------------------------*/
package org.deegree.graphics.transformation;

import org.deegree.model.spatialschema.Envelope;
import org.deegree.model.spatialschema.Position;

/**
 * <code>GeoTransformInterface</code> declares the methods which have to be implemented by each
 * class that executes a geographical coordinat transformation.
 *
 *
 * @author <a href="mailto:poth@lat-lon.de">Andreas Poth</a>
 * @author last edited by: $Author$
 *
 * @version $Revision$, $Date$
 */
public interface GeoTransform {
    /**
     *
     *
     * @param xdest
     *
     * @return source x coordinate
     */
    public double getSourceX( double xdest );

    /**
     *
     *
     * @param xsource
     *
     * @return destination x coordinate
     */
    public double getDestX( double xsource );

    /**
     *
     *
     * @param ydest
     *
     * @return source y coordinate
     */
    public double getSourceY( double ydest );

    /**
     *
     *
     * @param ysource
     *
     * @return destination y coordinate
     */
    public double getDestY( double ysource );

    /**
     * @param rect
     *
     */
    public void setSourceRect( Envelope rect );

    /**
     *
     *
     * @param xMin
     * @param yMin
     * @param xMax
     * @param yMax
     */
    public void setSourceRect( double xMin, double yMin, double xMax, double yMax );

    /**
     * @return source rectange
     *
     */
    public Envelope getSourceRect();

    /**
     * @param rect
     *
     */
    public void setDestRect( Envelope rect );

    /**
     *
     *
     * @param xMin
     * @param yMin
     * @param xMax
     * @param yMax
     */
    public void setDestRect( double xMin, double yMin, double xMax, double yMax );

    /**
     * @return destination rectange
     *
     */
    public Envelope getDestRect();

    /**
     *
     *
     * @param point
     *
     * @return source position
     */
    public Position getSourcePoint( Position point );

    /**
     *
     *
     * @param point
     *
     * @return destination position
     */
    public Position getDestPoint( Position point );
}
